File: transports/ruby/lib/csilgen/transport/cbor.rb
# frozen_string_literal: true

module Csilgen
  module Transport
    # Minimal canonical CBOR codec (RFC 8949), hand-rolled with zero gem deps so the
    # transport stays offline-testable. It supports exactly what CSIL envelopes need --
    # unsigned ints, negative ints, byte strings, text strings, arrays, maps, and tag
    # 24 -- and nothing else. Maps use core deterministic encoding (entries sorted by
    # the bytewise order of their *encoded* keys) so bytes match the conformance
    # vectors byte for byte across every language's reference library.
    module CBOR
      # Raised for malformed input or any construct outside the supported subset. A
      # subclass of the transport Error so a host can rescue the whole family at once.
      class CBORError < Error; end

      # A CBOR semantic tag wrapping one data item (RFC 8949 3.4). Data gives us the
      # structural == the conformance decode checks rely on for free.
      Tag = ::Data.define(:tag, :value)

      module_function

      # The signed-64 floor expressed as a negative-int argument ceiling. A major-1
      # argument above this names a value below i64::MIN; Go's reference rejects it
      # rather than letting the magnitude wrap, and we mirror that discipline even
      # though Ruby integers would not overflow, so behavior is identical cross-language.
      MAX_NINT_ARG = (1 << 63) - 1

      # Largest argument any CBOR head can carry; a larger Ruby Integer would make
      # pack("Q>") raise RangeError, which we convert into a clean CBORError instead.
      MAX_ARG = (1 << 64) - 1

      def encode(value)
        out = +"".b
        encode_into(out, value)
        out
      end

      def encode_into(out, value)
        case value
        when Tag
          encode_head(out, 6, value.tag)
          encode_into(out, value.value)
        when ::Integer
          if value.negative?
            # CBOR negative ints encode -1-n; the head argument is the magnitude minus one.
            encode_head(out, 1, -1 - value)
          else
            encode_head(out, 0, value)
          end
        when ::String
          # Binary-encoded strings are opaque bytes (major 2); anything else is text
          # (major 3). Distinguishing by encoding is the natural Ruby idiom and keeps
          # payload bytes from being re-validated as UTF-8.
          if value.encoding == ::Encoding::BINARY
            encode_head(out, 2, value.bytesize)
            out << value
          else
            encoded = value.encode("UTF-8")
            encode_head(out, 3, encoded.bytesize)
            out << encoded.b
          end
        when ::Array
          encode_head(out, 4, value.length)
          value.each { |item| encode_into(out, item) }
        when ::Hash
          # Deterministic map: sort entries by their encoded-key bytes. Because the key
          # head byte carries the length, shorter keys sort first -- the same
          # length-then-bytewise order RFC 8949 4.2.1 mandates. Keys are forced binary
          # so String#<=> compares bytewise rather than by the source encoding.
          items = value.map { |k, v| [encode(k), v] }
          items.sort_by! { |k, _| k }
          encode_head(out, 5, items.length)
          items.each do |key_bytes, v|
            out << key_bytes
            encode_into(out, v)
          end
        when true, false
          raise CBORError, "booleans are not part of any CSIL transport envelope"
        else
          raise CBORError, "cannot encode value of type #{value.class}"
        end
      end

      def encode_head(out, major, arg)
        raise CBORError, "CBOR argument #{arg} exceeds 64-bit range" if arg > MAX_ARG

        mt = major << 5
        if arg < 24
          out << [mt | arg].pack("C")
        elsif arg < 0x100
          out << [mt | 24, arg].pack("CC")
        elsif arg < 0x1_0000
          out << [mt | 25].pack("C") << [arg].pack("n")
        elsif arg < 0x1_0000_0000
          out << [mt | 26].pack("C") << [arg].pack("N")
        else
          out << [mt | 27].pack("C") << [arg].pack("Q>")
        end
      end

      # Decode exactly one CBOR item and reject trailing bytes: a CSIL envelope is a
      # single self-contained item, so leftover bytes are a malformed frame.
      def decode(data)
        data = data.b
        value, pos = decode_at(data, 0)
        raise CBORError, "trailing bytes after CBOR item" if pos != data.bytesize

        value
      end

      def decode_at(data, pos)
        need(data, pos, 1)
        initial = data.getbyte(pos)
        pos += 1
        major = initial >> 5
        info = initial & 0x1F
        arg, pos = read_argument(data, pos, info)

        case major
        when 0
          [arg, pos]
        when 1
          raise CBORError, "negative integer out of signed-64 range" if arg > MAX_NINT_ARG

          [-1 - arg, pos]
        when 2
          need(data, pos, arg)
          [data.byteslice(pos, arg), pos + arg]
        when 3
          need(data, pos, arg)
          text = data.byteslice(pos, arg).force_encoding("UTF-8")
          raise CBORError, "invalid UTF-8 in text string" unless text.valid_encoding?

          [text, pos + arg]
        when 4
          items = []
          arg.times do
            item, pos = decode_at(data, pos)
            items << item
          end
          [items, pos]
        when 5
          result = {}
          arg.times do
            key, pos = decode_at(data, pos)
            val, pos = decode_at(data, pos)
            result[key] = val
          end
          [result, pos]
        when 6
          inner, pos = decode_at(data, pos)
          [Tag.new(tag: arg, value: inner), pos]
        else
          raise CBORError, "unsupported CBOR major type #{major}"
        end
      end

      def read_argument(data, pos, info)
        if info < 24
          [info, pos]
        elsif info == 24
          need(data, pos, 1)
          [data.getbyte(pos), pos + 1]
        elsif info == 25
          need(data, pos, 2)
          [data.byteslice(pos, 2).unpack1("n"), pos + 2]
        elsif info == 26
          need(data, pos, 4)
          [data.byteslice(pos, 4).unpack1("N"), pos + 4]
        elsif info == 27
          need(data, pos, 8)
          [data.byteslice(pos, 8).unpack1("Q>"), pos + 8]
        else
          # 28..30 are reserved and 31 is indefinite-length, both forbidden in a CSIL envelope.
          raise CBORError, "unsupported additional-info value #{info}"
        end
      end

      def need(data, pos, n)
        raise CBORError, "unexpected end of CBOR input" if pos + n > data.bytesize
      end
    end
  end
end

File: transports/elixir/lib/csilgen/transport/cbor.ex
defmodule Csilgen.Transport.CBOR do
  @moduledoc """
  Minimal canonical CBOR codec (RFC 8949), hand-written and dependency-free.

  It supports exactly what the CSIL envelopes need — unsigned ints, negative ints,
  text strings, byte strings, arrays, maps, and tags — and nothing else. Maps use
  core deterministic encoding: entries are sorted by the bytewise lexicographic
  order of their *encoded* keys, which for Elixir binaries is the default `<=/2`,
  so the bytes match the conformance vectors regardless of insertion order.

  In-memory model (the items encode/decode produce and consume):

    * unsigned/negative int  → Elixir `integer()` (bignum: a CBOR uint64 round-trips
      with no overflow guard, and a negative-int decode needs no signed-floor guard)
    * text string            → bare `binary()`
    * byte string            → `{:bytes, binary()}` (tagged so the encoder never
      emits a tstr where a bstr is required — the bstr/tstr ambiguity trap)
    * array                  → `list()`
    * map                    → `{:map, [{key, value}]}` on encode; `%{}` on decode
    * tag                    → `{:tag, non_neg_integer(), value}`
  """

  @doc "Encodes one value to canonical CBOR bytes."
  @spec encode(term()) :: binary()
  def encode(value), do: value |> encode_iodata() |> IO.iodata_to_binary()

  # An envelope is a single self-contained item; trailing bytes are a malformed
  # frame (conventions doc §1), rejected here rather than silently ignored.
  @doc "Decodes one complete CBOR item, rejecting any trailing bytes."
  @spec decode(binary()) :: {:ok, term()} | {:error, atom()}
  def decode(bin) when is_binary(bin) do
    with {:ok, value, rest} <- decode_value(bin, 0) do
      if rest == <<>>, do: {:ok, value}, else: {:error, :trailing_bytes}
    end
  end

  # --- encode ---------------------------------------------------------------

  defp encode_iodata(n) when is_integer(n) and n >= 0, do: head(0, n)
  # CBOR negative ints encode -1-n; the argument is the magnitude minus one.
  defp encode_iodata(n) when is_integer(n) and n < 0, do: head(1, -1 - n)
  defp encode_iodata({:bytes, b}) when is_binary(b), do: [head(2, byte_size(b)), b]
  # UTF-8 is the native binary encoding, so byte_size/1 gives the wire length.
  defp encode_iodata(s) when is_binary(s), do: [head(3, byte_size(s)), s]

  defp encode_iodata(list) when is_list(list) do
    [head(4, length(list)) | Enum.map(list, &encode_iodata/1)]
  end

  defp encode_iodata({:map, entries}) when is_list(entries) do
    body =
      entries
      |> Enum.map(fn {k, v} -> {IO.iodata_to_binary(encode_iodata(k)), encode_iodata(v)} end)
      # Default <=/2 on binaries is bytewise lexicographic, shorter-prefix-first —
      # exactly RFC 8949 §4.2.1 once length is folded into the head byte.
      |> Enum.sort_by(fn {kb, _} -> kb end)
      |> Enum.flat_map(fn {kb, ve} -> [kb, ve] end)

    [head(5, length(entries)) | body]
  end

  defp encode_iodata({:tag, num, value}) when is_integer(num) and num >= 0 do
    [head(6, num), encode_iodata(value)]
  end

  # `<<n::16/32/64>>` default to big-endian unsigned — CBOR's network order — so no
  # qualifier is required, though the explicit form is kept for self-documentation.
  defp head(major, n) when n < 24, do: <<major::3, n::5>>
  defp head(major, n) when n < 0x100, do: <<major::3, 24::5, n::big-unsigned-integer-size(8)>>
  defp head(major, n) when n < 0x10000, do: <<major::3, 25::5, n::big-unsigned-integer-size(16)>>

  defp head(major, n) when n < 0x100000000,
    do: <<major::3, 26::5, n::big-unsigned-integer-size(32)>>

  defp head(major, n), do: <<major::3, 27::5, n::big-unsigned-integer-size(64)>>

  # --- decode ---------------------------------------------------------------

  defp decode_value(_bin, depth) when depth > 64, do: {:error, :nesting_limit_exceeded}

  defp decode_value(<<major::3, info::5, rest::binary>>, depth) do
    with {:ok, arg, rest} <- read_arg(info, rest) do
      decode_item(major, arg, rest, depth)
    end
  end

  defp decode_value(<<>>, _depth), do: {:error, :empty_input}

  defp read_arg(info, rest) when info < 24, do: {:ok, info, rest}
  defp read_arg(24, <<n::8, rest::binary>>), do: {:ok, n, rest}
  defp read_arg(25, <<n::16, rest::binary>>), do: {:ok, n, rest}
  defp read_arg(26, <<n::32, rest::binary>>), do: {:ok, n, rest}
  defp read_arg(27, <<n::64, rest::binary>>), do: {:ok, n, rest}
  defp read_arg(info, _rest) when info in 24..27, do: {:error, :truncated_argument}
  # 28..30 (reserved) and 31 (indefinite) are forbidden in CSIL envelopes.
  defp read_arg(_info, _rest), do: {:error, :indefinite_or_reserved}

  defp decode_item(0, arg, rest, _depth), do: {:ok, arg, rest}
  # Bignum integers make -1-arg exact with no signed-64 floor guard needed.
  defp decode_item(1, arg, rest, _depth), do: {:ok, -1 - arg, rest}

  defp decode_item(2, arg, rest, _depth) do
    # Pattern-matching a fixed-size slice never over-allocates on a forged length:
    # it only matches when the bytes are actually present.
    case rest do
      <<b::binary-size(^arg), rest2::binary>> -> {:ok, {:bytes, b}, rest2}
      _ -> {:error, :truncated_byte_string}
    end
  end

  defp decode_item(3, arg, rest, _depth) do
    case rest do
      <<s::binary-size(^arg), rest2::binary>> ->
        if String.valid?(s), do: {:ok, s, rest2}, else: {:error, :invalid_utf8}
      _ -> {:error, :truncated_text_string}
    end
  end

  defp decode_item(4, arg, rest, depth) when arg <= byte_size(rest),
    do: decode_seq(arg, rest, [], depth + 1)

  defp decode_item(4, _arg, _rest, _depth), do: {:error, :array_length_exceeds_remaining_input}

  defp decode_item(5, arg, rest, depth) when arg <= byte_size(rest),
    do: decode_map(arg, rest, [], depth + 1)

  defp decode_item(5, _arg, _rest, _depth), do: {:error, :map_length_exceeds_remaining_input}

  defp decode_item(6, arg, rest, depth) do
    with {:ok, value, rest2} <- decode_value(rest, depth + 1), do: {:ok, {:tag, arg, value}, rest2}
  end

  defp decode_item(7, _arg, _rest, _depth), do: {:error, :unsupported_simple_or_float}

  defp decode_seq(0, rest, acc, _depth), do: {:ok, Enum.reverse(acc), rest}

  defp decode_seq(n, rest, acc, depth) do
    with {:ok, item, rest2} <- decode_value(rest, depth) do
      decode_seq(n - 1, rest2, [item | acc], depth)
    end
  end

  defp decode_map(0, rest, acc, _depth), do: {:ok, Map.new(acc), rest}

  defp decode_map(n, rest, acc, depth) do
    with {:ok, key, rest2} <- decode_value(rest, depth),
         {:ok, value, rest3} <- decode_value(rest2, depth) do
      decode_map(n - 1, rest3, [{key, value} | acc], depth)
    end
  end
end
